### PR TITLE
Filter the source model upfront when there are less sites

### PR DESCRIPTION
**Pull request by micheles**
Disaggregation calculations produce not enough tasks (for instance the recent on for Brasil produced 57 tasks instead of the wanted 512). The reason is that the splitting in tasks is based on the total weight of the source model, while we should considered only the filtered weight (i.e. the discard the weight of the sources outside the integration distance).
This is the companion of https://github.com/gem/oq-hazardlib/pull/491/
The tests are green: https://ci.openquake.org/job/zdevel_oq-engine/2203/

### CODE REVIEW

File: openquake/commonlib/source.py [hunk 1]
```diff
@@ -616,7 +593,7 @@ def __init__(self, gsim_lt, source_model_lt, source_models,
         self.gsim_lt = gsim_lt
         self.source_model_lt = source_model_lt
         self.source_models = source_models
-        self.source_info = ()  # set by the SourceFilterSplitter
+        self.source_info = ()  # set by the SourceSitesFilterSplitter
```
Comment (daniviga) [inline openquake/commonlib/source.py]: Check the comment

File: openquake/commonlib/source.py [hunk 2]
```diff
@@ -626,6 +603,37 @@ def __init__(self, gsim_lt, source_model_lt, source_models,
             self.source_model_lt.num_samples,
             [sm.get_skeleton() for sm in self.source_models])
 
+    def filter(self, sitecol, src_sites_filter):
```
Comment (daniviga) [inline openquake/commonlib/source.py]: Maybe should be good to use `ss_filter` everywhere

File: openquake/commonlib/source.py [hunk 3]
```diff
@@ -767,9 +775,8 @@ def split_filter(src, sites, max_dist, random_seed):
             ss.serial = src.serial[start:start + nr]
             start += nr
         ss.id = src.id
-        with context(ss):
-            s = ss.filter_sites_by_distance_to_source(max_dist, sites)
-        if s is not None:
+        ss_sites = ss_filter.affected(ss, sites)
```
Comment (daniviga) [inline openquake/commonlib/source.py]: ss -> sub_source otherwise it makes confusion

File: openquake/commonlib/source.py [hunk 4]
```diff
@@ -880,20 +884,13 @@ def _gen_args(self, src_data, tiles):
...
-            for src in sources:
-                if self.filter_sources:
-                    with filter_mon, context(src):
-                        ok = src in tile
-                else:  # accept all sources
-                    ok = True
-                if ok:
-                    data.append((src, i, [], filter_mon.dt, 0))
+            for src, sites in self.ss_filter(sources, tile):
+                data.append((src, i, [], filter_mon.dt, 0))
```
Comment (daniviga) [inline openquake/commonlib/source.py]: You have lost the monitoring here

**Comment by daniviga** [comment]
Please update the changelog

**Comment by daniviga** [comment]
LGTM

File: openquake/commonlib/source.py (hunk 1 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Removed

File: openquake/commonlib/source.py (hunk 2 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Done

File: openquake/commonlib/source.py (hunk 3 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Used "split" for the split source

File: openquake/commonlib/source.py (hunk 4 above)
Comment (micheles) [inline openquake/commonlib/source.py]: Restored

**Pull request merged**